binaries/apiserver/load_test/load_tester.go

```diff
@@ -0,0 +1,409 @@
+	// cli args
+	action            string // the test's action (upload, download, both)
+	minDataSetSize    int    // the minimum data set size to use during the test
+	maxDataSetSize    int    // the maximium data set size to use during the test
+	numConcurrentActs int    // the number of concurrent upload/downloads we want to trigger
+	freq              int    // the frequency to repeat the test (0 = run once)
+	totalTime         int    // the total elapsed time to allow repeating tests to run
+	casGrpcAddr       string // the cas addr (<hostname>:<port>)
+
+	// data set fields
+	data                []byte   // the raw data set all uploads are derived from
+	dataSizes           []int    // the list of data sizes to use on the test
+	initUploadDigestIds []string // info on uploaded data sets for download testing
+	completedCnt        int      // track number of completed test actions
+	iterCnt             int      // track number of tests run (for freq != 0)
+	killRequested       bool     // will be set to true if/when a kill is requested
+	killReqeuestedMu	sync.RWMutex
+	status              StatusCode
+	// externals
+	dialer *dialer.ConstantResolver
+	stat   stats.StatsReceiver
+
+	// channels for goroutine coordination
+	stopTestIterations chan bool
+}
+
+type Args struct {
+	LogLevel    string
+	Action      string
+	DataSizeMin int
+	DataSizeMax int
+	NumTimes    int
+	Freq        int
+	TotalTime   int
+	CasGrpcAddr string
+}
+
+func MakeApiserverLoadTester(a *Args) *ApiserverLoadTester {
+	lt := ApiserverLoadTester{
+		action:            a.Action,
+		minDataSetSize:    a.DataSizeMin,
+		maxDataSetSize:    a.DataSizeMax,
+		numConcurrentActs: a.NumTimes,
+		freq:              a.Freq,
+		totalTime:         a.TotalTime,
+		casGrpcAddr:       a.CasGrpcAddr,
+		killRequested:     false,
+		status:            WaitingToStart,
+		iterCnt:           0,
+	}
+
+	// get the cas connection
+	lt.dialer = dialer.NewConstantResolver(lt.casGrpcAddr)
+
+	// initialize thes stats
+	statsReceiver, _ := stats.NewCustomStatsReceiver(stats.NewFinagleStatsRegistry, 0)
+	lt.stat = statsReceiver.Scope("cas_streaming")
+
+	lt.stopTestIterations = make(chan bool)
+	return &lt
+}
+
+/*
+Start the load test
+*/
+func (lt *ApiserverLoadTester) RunLoadTest() error {
+	/*
+	if its a repeating test (freq > 0) run the first iteration of the test in this process, start a goroutine
+	timing the frequency, then in this process loop running the next iterations when a timer sends a signal on its
+	channel, or stopping when time is up or a kill request is received
+	 */
+	lt.status = Initializing
+
+	// initialize the data sizes and data set for the test
+	err := lt.initTestData()
+	if err != nil {
+		return fmt.Errorf("couldn't initialize the test data:%s", err.Error())
+	}
+
+	if lt.getKillRequested() {
+		lt.drainStopChAndResetStatus()
+		return nil
+	}
+
+	if lt.freq == 0 {
+		lt.runOneIteration() // run the test once
+		lt.drainStopChAndResetStatus()
```

Comment:
nit - prefer to return in this block so the else block below can be removed, and that code is indented less